src/core/abr/learn_2_adapt.ts

```diff
@@ -0,0 +1,260 @@
+import { Observable } from "rxjs";
+import { Representation } from "../../manifest";
+import { IBeginRequestValue, IRequestInfo } from "./pending_requests_store";
+import { IABRMetricsEventValue, IABRStreamEvents } from "./representation_estimator";
+
+enum L2A_State {
+  NoChange, // Only 1 bitrate available
+  Start,
+  Steady,
+}
+
+interface IRulesContext {
+  qualityBasedBitrate: number;
+  bufferGap: number;
+  lastRequestInfo: IRequestInfo;
+  speed: number;
+}
+
+interface L2AParameters {
+  w: number[];
+  prevW: number[];
+  Q: number;
+  segmentRequestStart: number;
+  segmentDownloadFinish: number;
+  bufferTargetLvl: number;
+}
+
+interface ICxtArg {
+  bitrates: number[];
+  representations: Representation[];
+  streamEvents$: Observable<IABRStreamEvents>;
+}
+
+class L2ARule {
```

Comment:
In a more Rx-y way all the methods but one could just listen on an Observable. Does it makes sense to convert it to a function? I do not know when `getMaxPossibleRepresentation` will be called, that's the missing piece for me.